### PR TITLE
Make bufferowner controlled by himself

### PR DESCRIPTION
**Pull request by SergeyMokin**
Added test case when it thrown an exception.
We can just add one more if check, by I prefer to see that BufferOwner control buffer by himself how much it possible.
Just one more think to move grow control to BufferOwner and not share buffer at all.

### CODE REVIEW

File: src/ListPool/ListPool.cs [hunk 1]
```diff
@@ -50,7 +49,6 @@ public ListPool(IEnumerable<TSource> source)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Add(TSource item)
         {
-            if (!_bufferOwner.IsValid) _bufferOwner = new BufferOwner<TSource>(MinimumCapacity);
             if (_itemsCount >= _bufferOwner.Buffer.Length) _bufferOwner.GrowDoubleSize();
```
Comment (SergeyMokin) [inline src/ListPool/ListPool.cs]: I mentioned about it, need to think is it a good idea or not to move control of buffer to bufferowner

File: src/ListPool/ListPool.cs [hunk 2]
```diff
@@ -123,7 +120,7 @@ public void RemoveAt(int index)
         [Pure]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public readonly Enumerator<TSource> GetEnumerator() =>
-            new Enumerator<TSource>(in _bufferOwner.Buffer, in _itemsCount);
+            new Enumerator<TSource>(_bufferOwner.Buffer, in _itemsCount);
```
Comment (SergeyMokin) [inline src/ListPool/ListPool.cs]: and here share BufferOwner (not Buffer)

**Comment by faustodavid** [comment]
Nice catch.! Maybe is better to create an issue with the **Contains**.  Because we want to be fast, we should first consider the implications of having a new "if" statement in the hot path "get buffer". 

It is shitty that for the nature of struct we don't have constructors and we need all these boilerplate code. I am considering opening an issue for this repository where we state that the only way to create a valid ListPool is by an extension method, so we don't need all the checks in hot paths and we can resemble better the ArrayPool.

using var numbers = ListPool<int>.Shared.Rent(100);

But we cant restricted if the developer does

using var numbers = ListPool<int>()

But then we can have a valid flag for ListPool itself. 

The downside of this approach is for the serializer, which is another high priority issue that we should address, it will be harder to do.

**Comment by SergeyMokin** [comment]
won't do like this

**Pull request closed without merging**